Outbound shipment auto allocation

mark-prins opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #2227, #2266, #2264, #1983, #2255

# 👩🏻‍💻 What does this PR do? 
 <!-- Explain the changes you made, and why they're needed. Add a screenshot if you've made any UI changes!  -->
See #2261 for the discussion and requested changes. Reinstates the auto allocation which was removed in #1189

- Allow saving of lines with 0 quantity
- Show a warning if you click OK when the quantity is 0; this asks you to click OK again to confirm
- Reinstate the auto allocation, this happens when the issue quantity is changed or the pack size is changed
- Prune 0 quantity lines when changing status from NEW
- Show a warning when changing status
- Fix the strange warning issue - where the notification shows incorrectly
- Show the warning when the pack size prevents the requested quantity being issued
- Show a warning when a placeholder has been allocated due to insufficient stock


<img width="1060" alt="Screenshot 2023-09-27 at 11 39 12 AM" src="https://github.com/openmsupply/open-msupply/assets/9192912/d7b8c100-8776-425f-b722-16ac239af774">

<img width="1056" alt="Screenshot 2023-09-27 at 11 38 44 AM" src="https://github.com/openmsupply/open-msupply/assets/9192912/c2bf8813-36dd-43b3-89d7-2ce23b818b9f">

![Screenshot 2023-09-27 at 10 31 44 PM](https://github.com/openmsupply/open-msupply/assets/9192912/023e1337-feda-45d4-a09d-a235ace30355)

![Screenshot 2023-09-27 at 7 10 23 PM](https://github.com/openmsupply/open-msupply/assets/9192912/7a3f870b-64ed-4de4-be78-0a2577531716)




# 🧪 How has/should this change been tested? 
<!-- Explain how to setup for testing here if it is not already obvious, and how you've tested this PR. -->

- Create a new outbound shipment and try to add an item with 0 quantity
- Update an existing line to 0  and check the messaging
- Try the auto allocation
- Try the two way updates - editing a line and checking that the issue quantity updates
- Change status and check that lines are pruned


## 💌 Any notes for the reviewer?
<!-- eg. Do you have any specific questions for the reviewer? Is there a high risk/complicated change they should focus on? If there are any general areas of the codebase your changes might have have touched or could cause side effects to, mention them here.-->

## 📃 Documentation
<!-- Note down any areas which require documentation updates -->
- [ ]  Auto allocation
- [ ] UI changes

mark-prins commented:
Just draft at this stage, to get feedback. Haven't done prescriptions yet.

roxy-dao commented:
Looks good to me so far ~, thanks Mark~

imo, the `placeholder-allocated` isn't clear enough... and there are these two issues to address this #2255 and #1983 🤔

CarlosNZ commented:
Cool, looking good. Here's just a few observations (in no particular order) from my initial testing:

- Maybe the auto-allocation should only happen when losing focus rather than on every keystroke of the quantity input? I guess your reasoning is that you probably want to go straight from quantity entry to "OK" and you should see the allocations before clicking OK? I guess that makes sense, but perhaps a short debounce then on the input? (It's particularly noticeable when you add a quantity that causes the "Due to the pack sizes available a total quantity ..." message to appear -- when typing a number, you get the warning after the first keystroke which then disappears, but it shouldn't really appear until you've finished typing you're intended number -- although maybe this is a separate issue?)
- Shouldn't the "OK & Next" button behave the same as the "OK" button in terms of the zero-quantity warning? Currently it just goes to the next with no feedback.
- If we add a quantity, but then go back in to that line and change the quantity to 0, we get the warning and the 0 allocation display, but when clicking "OK" the line in the table hasn't been updated (to 0) -- still shows previous quantity, with "black" formatting.
- This is pretty nitpicky, but I feel like the top/bottom margins around the warning boxes could be a little wider? 
![Screenshot 2023-10-03 at 11 54 04 AM](https://github.com/openmsupply/open-msupply/assets/5456533/62a10643-568d-4378-945e-9a1f49f05d3e)  
Maybe somewhere between where it is currently and the margin below the "Issue" line?
- There's an error that happens in some situations, though I'm not 100% clear what it is. Here's how I can reproduce it on my setup:
  - Open a line and add a quantity higher than available. You'll get the "Not enough stock" warning. 
  - Close the item
  - Open another item and input a quantity
  - Even when this quantity is valid, the "OK" button won't work (and sometimes an error snackbar appears)
  - I suspect the warning from the previous item is still "active" somehow, so we probably need to explicitly clear it when closing the modal

I'll have a closer look at the code once this is no longer in Draft. Working nicely though and I like the warning when 0 solution.

mark-prins commented:
Thanks for the comments @roxy-dao & @CarlosNZ : super helpful!!

> Maybe the auto-allocation should only happen when losing focus..

I did actually try both the `onBlur` and a debounce - and missed having the immediate feedback between what I type and the changes that happen with the stock line allocation.

Even having a debounce made it harder to get an intuitive sense that what I've entered related to the stock lines. Having the notifications pop in was annoying but I thought the trade off worth it. I'll try adding a small debounce to the notification display and see how that feels 🙏

mark-prins commented:
Updated with review feedback and have added to prescriptions.
After implementing the more flexible `Alert` component, I've removed `InfoPanel` and replaced usages, and added a story.

The prescriptions now allow adding of `0` quantity rows, if the prescription is `New` only. If you add when `Picked` the rows are removed automatically. The rows are also pruned when you changed to `Verified`. I thought that this is the best compromise for the upcoming change to always have prescriptions `Picked` when items are added.. though of course things become a little weird. I don't expect that anyone would add a 0 quantity prescription line, so this is a very niche scenario anyway. 

Couldn't reproduce your last point @CarlosNZ and so I don't know if that's resolved or data specific. Can you raise a new issue if that's a valid problem still.

<img width="1062" alt="Screenshot 2023-10-04 at 5 42 50 PM" src="https://github.com/openmsupply/open-msupply/assets/9192912/c47e0842-a8e7-4709-89fe-39b83fd99328">
<img width="1057" alt="Screenshot 2023-10-04 at 10 14 42 PM" src="https://github.com/openmsupply/open-msupply/assets/9192912/56f25128-953a-4f80-aba0-c9e9abd262df">
<img width="1030" alt="Screenshot 2023-10-05 at 8 19 22 AM" src="https://github.com/openmsupply/open-msupply/assets/9192912/cffee997-51e3-4681-acfc-861c418d2dbe">
<img width="863" alt="Screenshot 2023-10-05 at 8 21 13 AM" src="https://github.com/openmsupply/open-msupply/assets/9192912/7c5fc5e9-b1a2-4fed-81c6-1ef4f8b1b9e1">

CarlosNZ commented:
Hi @mark-prins -- I'm getting server compilation errors on this branch now (which are not on develop).

The console produces screeds of text, but here's the last bit if that helps:
![Screenshot 2023-10-05 at 1 16 40 PM](https://github.com/openmsupply/open-msupply/assets/5456533/82682d33-374b-48b5-a25c-cf550ef6f4f6)

roxy-dao commented:
> Hi @mark-prins -- I'm getting server compilation errors on this branch now (which are not on develop).
> 
try `cargo clean`

CarlosNZ commented:
> > Hi @mark-prins -- I'm getting server compilation errors on this branch now (which are not on develop).
> 
> try `cargo clean`

Thanks -- sorted :)

CarlosNZ reviewed (commented):
Looking good, thanks. I've made one little comment about the message persisting on "OK and Next".

Couple of overall thoughts:
- I think I preferred the previous iteration when the notification boxes weren't full width. They seem to have too much empty space on the right now.
- It seems a bit odd to have the message "Some stock lines are expired and cannot be allocated" not appear till after you've allocated something. This info is known from the get-go, so it feels like that message should already be there when you open the modal. Would that be hard to do? (The other messages make sense when they appear)
- I can't reproduce the bug I mentioned in the last review any more, so whatever you've done seems to have fixed it 👍

mark-prins commented:
> I think I preferred the previous iteration when the notification boxes weren't full width. They seem to have too much empty space on the right now.

oh. unintentional - will see what change did that.

mark-prins commented on client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx:
```diff
@@ -159,6 +152,56 @@ export const OutboundLineEdit: React.FC<ItemDetailsModalProps> = ({
+      await onSave();
+      setIsDirty(false);
+      if (!!placeholder) {
+        const infoSnack = info(t('message.placeholder-line'));
+        infoSnack();
```
no reason that I can see other than making it clearer. I've had people review and insist that it be split out, iirc that was just about clarity. In this case - the lines are pre-exisiting, simply moved from elsewhere in the file.

mark-prins commented:
> It seems a bit odd to have the message...

I prefer having it tied to the auto allocation. I don't want to have an alert message saying that there's on hold / expired stock which shows all the time, particularly when the individual lines are already showing this (red font for the expiry date and the 'on hold' column). The intention of the message is to explain why the allocation behaved how it did, and I'd rather not clutter the UI with an additional message.

CarlosNZ reviewed (approved):
Okay, LGTM.
